src/runtime/types.cpp

```diff
@@ -2929,6 +2931,56 @@ extern "C" PyUnicodeObject* _PyUnicode_New(Py_ssize_t length) noexcept {
     return unicode;
 }
 
+// Normally we don't call the Python tp_ slots that are present to support
+// CPython's reference-counted garbage collection. However, for a few types
+// like weak references, we still rely on reference-counting to some extent.
+static void setTypeGCProxy(BoxedClass* cls) {
+    cls->tp_alloc = PystonType_GenericAlloc;
+    cls->gc_visit = proxy_to_tp_traverse;
+    cls->tp_dealloc = proxy_to_tp_clear;
```

Comment:
might be good to add a comment here "we can't use the normal tp_dealloc here, since proxy_dealloc explicitly frees self but doesn't go through tp_free, etc..." so we don't forget and ask the questions again :)